diff --git a/CMakeLists.txt b/CMakeLists.txt
index 8a03ff0..a411c8d 100755
--- a/CMakeLists.txt
+++ b/CMakeLists.txt
@@ -574,7 +574,7 @@ install(TARGETS butterflypack EXPORT MYBPACKTargets
   ARCHIVE DESTINATION ${CMAKE_INSTALL_LIBDIR}
   RUNTIME DESTINATION ${CMAKE_INSTALL_BINDIR})
 
-add_subdirectory(EXAMPLE)
+# add_subdirectory(EXAMPLE)
 
 
 # Documentation
diff --git a/PrecisionPreprocessing.sh b/PrecisionPreprocessing.sh
index 2659667..772960a 100755
--- a/PrecisionPreprocessing.sh
+++ b/PrecisionPreprocessing.sh
@@ -191,7 +191,7 @@ sed -i -e "s/py_bpack/z_py_bpack/g" $ZSRCDIR/CMakeLists.txt
 sed -i -e "s/c_bf_/z_c_bf_/g" $ZSRCDIR/*.h
 sed -i -e "s/c_bf_/z_c_bf_/g" $ZSRCDIR/*.cpp
 sed -i -e "s/BPACK_WRAP/z_BPACK_WRAP/g" $ZSRCDIR/*.h
-sed -i -e "s/BPACK_wrapper.h/zBPACK_wrapper.h/g" $ZSRCDIR/*.cpp
+sed -i -e "s/\"BPACK_wrapper.h\"/\"zBPACK_wrapper.h\"/g" $ZSRCDIR/*.cpp
 sed -i -e "s/parsec_hmat_ptg_bridge.h/zparsec_hmat_ptg_bridge.h/g" $ZSRCDIR/*.c
 sed -i -e "s/parsec_hmat_ptg_bridge.h/zparsec_hmat_ptg_bridge.h/g" $ZSRCDIR/*.jdf
 sed -i -e "s/c_bpack_/z_c_bpack_/g" $ZSRCDIR/*.f90
@@ -257,7 +257,7 @@ sed -i -e "s/py_bpack/d_py_bpack/g" $DSRCDIR/CMakeLists.txt
 sed -i -e "s/c_bf_/d_c_bf_/g" $DSRCDIR/*.h
 sed -i -e "s/c_bf_/d_c_bf_/g" $DSRCDIR/*.cpp
 sed -i -e "s/BPACK_WRAP/d_BPACK_WRAP/g" $DSRCDIR/*.h
-sed -i -e "s/BPACK_wrapper.h/dBPACK_wrapper.h/g" $DSRCDIR/*.cpp
+sed -i -e "s/\"BPACK_wrapper.h\"/\"dBPACK_wrapper.h\"/g" $DSRCDIR/*.cpp
 sed -i -e "s/parsec_hmat_ptg_bridge.h/dparsec_hmat_ptg_bridge.h/g" $DSRCDIR/*.c
 sed -i -e "s/parsec_hmat_ptg_bridge.h/dparsec_hmat_ptg_bridge.h/g" $DSRCDIR/*.jdf
 sed -i -e "s/c_bpack_/d_c_bpack_/g" $DSRCDIR/*.f90
@@ -322,7 +322,7 @@ sed -i -e "s/py_bpack/c_py_bpack/g" $CSRCDIR/CMakeLists.txt
 sed -i -e "s/c_bf_/c_c_bf_/g" $CSRCDIR/*.h
 sed -i -e "s/c_bf_/c_c_bf_/g" $CSRCDIR/*.cpp
 sed -i -e "s/BPACK_WRAP/c_BPACK_WRAP/g" $CSRCDIR/*.h
-sed -i -e "s/BPACK_wrapper.h/cBPACK_wrapper.h/g" $CSRCDIR/*.cpp
+sed -i -e "s/\"BPACK_wrapper.h\"/\"cBPACK_wrapper.h\"/g" $CSRCDIR/*.cpp
 sed -i -e "s/parsec_hmat_ptg_bridge.h/cparsec_hmat_ptg_bridge.h/g" $CSRCDIR/*.c
 sed -i -e "s/parsec_hmat_ptg_bridge.h/cparsec_hmat_ptg_bridge.h/g" $CSRCDIR/*.jdf
 sed -i -e "s/c_bpack_/c_c_bpack_/g" $CSRCDIR/*.f90
@@ -387,7 +387,7 @@ sed -i -e "s/py_bpack/s_py_bpack/g" $SSRCDIR/CMakeLists.txt
 sed -i -e "s/c_bf_/s_c_bf_/g" $SSRCDIR/*.h
 sed -i -e "s/c_bf_/s_c_bf_/g" $SSRCDIR/*.cpp
 sed -i -e "s/BPACK_WRAP/s_BPACK_WRAP/g" $SSRCDIR/*.h
-sed -i -e "s/BPACK_wrapper.h/sBPACK_wrapper.h/g" $SSRCDIR/*.cpp
+sed -i -e "s/\"BPACK_wrapper.h\"/\"sBPACK_wrapper.h\"/g" $SSRCDIR/*.cpp
 sed -i -e "s/parsec_hmat_ptg_bridge.h/sparsec_hmat_ptg_bridge.h/g" $SSRCDIR/*.c
 sed -i -e "s/parsec_hmat_ptg_bridge.h/sparsec_hmat_ptg_bridge.h/g" $SSRCDIR/*.jdf
 sed -i -e "s/c_bpack_/s_c_bpack_/g" $SSRCDIR/*.f90
diff --git a/SRC/BPACK_factor.f90 b/SRC/BPACK_factor.f90
index f4e12ff..2a911f8 100755
--- a/SRC/BPACK_factor.f90
+++ b/SRC/BPACK_factor.f90
@@ -1057,12 +1057,18 @@ endif
 #endif
 #endif
                     do ij = 1, (num_blocks-kk)*(num_blocks-kk)
+                        t_start = 0d0
+#ifdef HAVE_OPENMP
                         t_start = omp_get_wtime()
+#endif
                         j = (ij - 1)/(num_blocks-kk) + 1
                         i = mod(ij - 1, (num_blocks-kk)) + 1
                         i = i + kk
                         j = j + kk
+                        tid = 0
+#ifdef HAVE_OPENMP
                         tid = omp_get_thread_num()
+#endif
                         call g2l(i, num_blocks, nprow, 1, iproc1, myi1)
                         call g2l(j, num_blocks, npcol, 1, jproc1, myj1)
                         if(iproc1==myrow .and. jproc1==mycol)then
@@ -1079,7 +1085,10 @@ endif
                             call Hmat_block_delete(blocks_ll)
                             call Hmat_block_delete(blocks_uu)
                         endif
+                        t_stop = 0d0
+#ifdef HAVE_OPENMP
                         t_stop  = omp_get_wtime()
+#endif
                         t_total(tid+1) = t_total(tid+1) + (t_stop - t_start)
                     enddo
 #ifdef HAVE_TOPLEVEL_OPENMP
