diff --git a/mesh/mesh.cpp b/mesh/mesh.cpp
index a3f02efe71..0671228b8c 100644
--- a/mesh/mesh.cpp
+++ b/mesh/mesh.cpp
@@ -3029,6 +3029,7 @@ void Mesh::DoNodeReorder(DSTable *old_v_to_v, Table *old_elem_vert)
    const int num_edge_dofs = old_dofs.Size();
 
    // Save the original nodes
+   Nodes->HostReadWrite(); // for "(*Nodes)() = "
    const Vector onodes = *Nodes;
 
    // vertex dofs do not need to be moved
