diff --git a/mesh/mesh.cpp b/mesh/mesh.cpp
index 446257416..4af5bc785 100644
--- a/mesh/mesh.cpp
+++ b/mesh/mesh.cpp
@@ -7152,6 +7152,7 @@ int *Mesh::GeneratePartitioning(int nparts, int part_method)
       idx_t options[5];
 #else
       idx_t ncon = 1;
+      real_t kbval = 0.001;
       idx_t errflag;
       idx_t options[40];
 #endif
@@ -7236,7 +7237,7 @@ int *Mesh::GeneratePartitioning(int nparts, int part_method)
                                             NULL,
                                             &mparts,
                                             NULL,
-                                            NULL,
+                                            &kbval,
                                             options,
                                             &edgecut,
                                             mpartitioning);
@@ -7274,7 +7275,7 @@ int *Mesh::GeneratePartitioning(int nparts, int part_method)
                                        NULL,
                                        &mparts,
                                        NULL,
-                                       NULL,
+                                       &kbval,
                                        options,
                                        &edgecut,
                                        mpartitioning);
@@ -7313,7 +7314,7 @@ int *Mesh::GeneratePartitioning(int nparts, int part_method)
                                        NULL,
                                        &mparts,
                                        NULL,
-                                       NULL,
+                                       &kbval,
                                        options,
                                        &edgecut,
                                        mpartitioning);
