diff --git a/mesh/mesh.cpp b/mesh/mesh.cpp
index 8b32f1eba..e18def619 100644
--- a/mesh/mesh.cpp
+++ b/mesh/mesh.cpp
@@ -11811,7 +11811,7 @@ void Mesh::PrintVTU(std::ostream &os, int ref, VTKFormat format,
             const int *p = VTKGeometry::VertexPermutation[geom];
             for (int k = 0; k < nv; k++, j++)
             {
-               WriteBinaryOrASCII(os, buf, np + RG[p ? p[j] : j], " ",
+               WriteBinaryOrASCII(os, buf, np + RG[p ? (j - k + p[k]) : j], " ",
                                   format);
             }
             if (format == VTKFormat::ASCII) { os << '\n'; }
