diff --git a/mesh/mesh.cpp b/mesh/mesh.cpp
index d79ee99cf..660f299f0 100644
--- a/mesh/mesh.cpp
+++ b/mesh/mesh.cpp
@@ -9570,9 +9570,11 @@ double Mesh::AggregateError(const Array<double> &elem_error,
          case 0: error = std::min(error, err_fine); break;
          case 1: error += err_fine; break;
          case 2: error = std::max(error, err_fine); break;
+         case 3: error = std::pow(err_fine, 2.0); break;
+         default: MFEM_ABORT("Invalid operation.");
       }
    }
-   return error;
+   return (op == 3) ? std::sqrt(error) : error;
 }
 
 bool Mesh::NonconformingDerefinement(Array<double> &elem_error,
