diff --git a/mesh/pncmesh.cpp b/mesh/pncmesh.cpp
index 45433ee..caee26a 100644
--- a/mesh/pncmesh.cpp
+++ b/mesh/pncmesh.cpp
@@ -993,6 +993,7 @@ void ParNCMesh::GetConformingSharedStructures(ParMesh &pmesh)
 void ParNCMesh::GetFaceNeighbors(ParMesh &pmesh)
 {
    ClearAuxPM();
+   pmesh.DeleteFaceNbrData();

    const NCList &shared = (Dim == 3) ? GetSharedFaces() : GetSharedEdges();
    const NCList &full_list = (Dim == 3) ? GetFaceList() : GetEdgeList();
@@ -1421,6 +1422,8 @@ void ParNCMesh::GetFaceNeighbors(ParMesh &pmesh)
    }
    // NOTE: this function skips ParMesh::send_face_nbr_vertices and
    // ParMesh::face_nbr_vertices_offset, these are not used outside of ParMesh
+
+   pmesh.have_face_nbr_data = true;
 }

 void ParNCMesh::ClearAuxPM()
