diff --git a/fem/gridfunc.cpp b/fem/gridfunc.cpp
index 310d8d704..bd5504d63 100644
--- a/fem/gridfunc.cpp
+++ b/fem/gridfunc.cpp
@@ -1924,9 +1924,9 @@ void GridFunction::GetNodalValues(Vector &nval, int vdim) const
    Array<double> values;
    Array<int> overlap(fes->GetNV());
    nval.SetSize(fes->GetNV());
-
    nval = 0.0;
    overlap = 0;
+   nval.HostReadWrite();
    for (i = 0; i < fes->GetNE(); i++)
    {
       fes->GetElementVertices(i, vertices);
