diff --git a/Makefile b/Makefile
index 483c57f..d3286cd 100644
--- a/Makefile
+++ b/Makefile
@@ -166,7 +166,7 @@ PEDANTIC      ?=
 PEDANTICFLAGS ?= -Werror -pedantic

 # Compiler flags
-OPT    ?= -O $(MARCHFLAG) $(OPT.$(CC_VENDOR)) $(OMP_SIMD_FLAG)
+override OPT    = -ffp-contract=fast -fopenmp-simd
 CFLAGS ?= $(OPT) $(CFLAGS.$(CC_VENDOR)) $(if $(PEDANTIC),$(PEDANTICFLAGS))
 CXXFLAGS ?= $(OPT) $(CXXFLAGS.$(CC_VENDOR)) $(if $(PEDANTIC),$(PEDANTICFLAGS))
 FFLAGS ?= $(OPT) $(FFLAGS.$(FC_VENDOR))
