diff --git a/CMakeLists.txt b/CMakeLists.txt
index 83144b4..98c0cc2 100644
--- a/CMakeLists.txt
+++ b/CMakeLists.txt
@@ -41,8 +41,8 @@
 ##                                                        ##
 ############################################################
 
-project(SCOTCH)
 cmake_minimum_required(VERSION 3.10)
+project(SCOTCH)
 enable_language(C Fortran)
 
 # Add module directory
diff --git a/src/esmumps/CMakeLists.txt b/src/esmumps/CMakeLists.txt
index 62114b5..f3eedf1 100644
--- a/src/esmumps/CMakeLists.txt
+++ b/src/esmumps/CMakeLists.txt
@@ -106,7 +106,6 @@ target_include_directories(esmumps PRIVATE
   $<INSTALL_INTERFACE:include>)
 
 target_link_libraries(esmumps PRIVATE scotch)
-
 add_dependencies(esmumps scotch_h)
 
 add_executable(test_esmumps test_esmumps.c)
@@ -135,9 +134,12 @@ install(TARGETS esmumps
 
 # Dummy PTEsMUMPS library file
 add_library(ptesmumps ${esmumps_src} ${esmumps_headers})
-
 set_target_properties(ptesmumps PROPERTIES PUBLIC_HEADER "${GENERATED_INCLUDE_DIR}/esmumps.h")
 
+if(APPLE)
+  set_target_properties(ptesmumps PROPERTIES LINK_FLAGS "-undefined dynamic_lookup")
+endif(APPLE)
+
 target_include_directories(ptesmumps PRIVATE
   $<BUILD_INTERFACE:${LIBSCOTCH_INCLUDE_DIR}>
   $<BUILD_INTERFACE:${GENERATED_INCLUDE_DIR}>
diff --git a/src/libscotch/dummysizes.c b/src/libscotch/dummysizes.c
index fbd7b52..7455672 100644
--- a/src/libscotch/dummysizes.c
+++ b/src/libscotch/dummysizes.c
@@ -253,6 +253,8 @@ char *                      argv[])
   subsFill ("DUMMYPATCHLEVEL", EXPAND (SCOTCH_PATCHLEVEL_NUM));
   subsFill ("DUMMYSIZEBYTEIDX", subsSizeByte (sizeof (IDX)));
   subsFill ("DUMMYSIZEBYTENUM", subsSizeByte (sizeof (INT)));
+  subsFill ("DUMMYSIZEBITIDX", subsSizeByte (sizeof (IDX) * 8));
+  subsFill ("DUMMYSIZEBITNUM", subsSizeByte (sizeof (INT) * 8));
   subsFill ("DUMMYSIZEARCHDOM", subsSize (sizeof (ArchDom)));
   subsFill ("DUMMYSIZEARCH", subsSize (sizeof (Arch)));
   subsFill ("DUMMYSIZEGEOM", subsSize (sizeof (Geom)));
diff --git a/src/libscotchmetis/library_metis.h b/src/libscotchmetis/library_metis.h
index a259950..dd5f10a 100644
--- a/src/libscotchmetis/library_metis.h
+++ b/src/libscotchmetis/library_metis.h
@@ -63,6 +63,15 @@
 /**                                                        **/
 /************************************************************/
 
+/*
+**  The type and structure definitions.
+*/
+
+#ifndef SCOTCH_H                                  /* In case "scotch.h" not included before */
+#include <stdint.h>
+typedef DUMMYINT SCOTCH_Num;
+#endif /* SCOTCH_H */
+
 /*
 **  The defines.
 */
@@ -92,6 +101,7 @@
 #define SCOTCH_METIS_DATATYPES
 typedef SCOTCH_Num          idx_t;
 typedef double              real_t;
+#define IDXTYPEWIDTH        DUMMYSIZEBITNUM
 #endif /* SCOTCH_METIS_DATATYPES */
 
 #ifndef SCOTCH_METIS_OPTIONS
@@ -146,14 +156,6 @@ typedef enum {
 } rstatus_et;
 #endif /* SCOTCH_METIS_RETURN */
 
-/*
-**  The type and structure definitions.
-*/
-
-#ifndef SCOTCH_H                                  /* In case "scotch.h" not included before */
-typedef DUMMYINT SCOTCH_Num;
-#endif /* SCOTCH_H */
-
 /*
 **  The function prototypes.
 */
@@ -174,7 +176,7 @@ int                         SCOTCH_METIS_V5_PartGraphKway (const SCOTCH_Num * co
 int                         SCOTCH_METIS_V5_PartGraphRecursive (const SCOTCH_Num * const, const SCOTCH_Num * const, const SCOTCH_Num * const, const SCOTCH_Num * const, const SCOTCH_Num * const, const SCOTCH_Num * const, const SCOTCH_Num * const, const SCOTCH_Num * const, const double * const, const double * const, const SCOTCH_Num * const, SCOTCH_Num * const, SCOTCH_Num * const);
 
 #ifndef SCOTCH_METIS_VERSION
-#define SCOTCH_METIS_VERSION        3             /* MeTiS API version is 3 by default */
+#define SCOTCH_METIS_VERSION        5             /* MeTiS API version is 5 by default */
 #endif /* SCOTCH_METIS_VERSION */
 
 #if (SCOTCH_METIS_VERSION == 3)
diff --git a/src/libscotchmetis/library_parmetis.h b/src/libscotchmetis/library_parmetis.h
index 945819f..897823b 100644
--- a/src/libscotchmetis/library_parmetis.h
+++ b/src/libscotchmetis/library_parmetis.h
@@ -61,6 +61,15 @@
 /**                                                        **/
 /************************************************************/
 
+/*
+**  The type and structure definitions.
+*/
+
+#ifndef SCOTCH_H                                  /* In case "scotch.h" not included before */
+#include <stdint.h>
+typedef DUMMYINT SCOTCH_Num;
+#endif /* SCOTCH_H */
+
 /*
 **  The defines and includes.
 */
@@ -93,6 +102,13 @@
 #define METISNAME4(p,s)             p##s
 #endif /* METISNAMEL */
 
+#ifndef SCOTCH_METIS_DATATYPES
+#define SCOTCH_METIS_DATATYPES
+typedef SCOTCH_Num          idx_t;
+typedef double              real_t;
+#define IDXTYPEWIDTH        DUMMYSIZEBITNUM
+#endif /* SCOTCH_METIS_DATATYPES */
+
 #ifndef SCOTCH_METIS_RETURN
 #define SCOTCH_METIS_RETURN
 typedef enum {
@@ -103,14 +119,6 @@ typedef enum {
 } rstatus_et;
 #endif /* SCOTCH_METIS_RETURN */
 
-/*
-**  The type and structure definitions.
-*/
-
-#ifndef SCOTCH_H                                  /* In case "scotch.h" not included before */
-typedef DUMMYINT SCOTCH_Num;
-#endif /* SCOTCH_H */
-
 /*
 **  The function prototypes.
 */
diff --git a/src/libscotchmetis/metis_graph_part.c b/src/libscotchmetis/metis_graph_part.c
index 867c851..62c4610 100644
--- a/src/libscotchmetis/metis_graph_part.c
+++ b/src/libscotchmetis/metis_graph_part.c
@@ -548,16 +548,18 @@ SCOTCH_Num * const          objval,
 SCOTCH_Num * const          part)
 {
   SCOTCH_Num          baseval;
+  double              kbalval;
 
   baseval = ((options != NULL) && (options != xadj)) ? options[METIS_OPTION_NUMBERING] : 0;
+  kbalval = (ubvec != NULL) ? *ubvec : 0.01;
 
   return ((vsize == NULL)
           ? _SCOTCH_METIS_PartGraph (nvtxs, ncon, xadj, adjncy, vwgt, adjwgt,
                                      &baseval, nparts, tpwgts, options, objval, part,
-                                     SCOTCH_STRATDEFAULT, ubvec)
+                                     SCOTCH_STRATDEFAULT, &kbalval)
           : _SCOTCH_METIS_PartGraph_Volume (nvtxs, ncon, xadj, adjncy, vwgt, vsize,
                                             &baseval, nparts, tpwgts, options, objval, part,
-                                            SCOTCH_STRATDEFAULT, ubvec));
+                                            SCOTCH_STRATDEFAULT, &kbalval));
 }
 
 /*
@@ -581,16 +583,18 @@ SCOTCH_Num * const          objval,
 SCOTCH_Num * const          part)
 {
   SCOTCH_Num          baseval;
+  double              kbalval;
 
   baseval = ((options != NULL) && (options != xadj)) ? options[METIS_OPTION_NUMBERING] : 0;
+  kbalval = (ubvec != NULL) ? *ubvec : 0.01;
 
   return ((vsize == NULL)
           ? _SCOTCH_METIS_PartGraph (nvtxs, ncon, xadj, adjncy, vwgt, adjwgt,
                                      &baseval, nparts, tpwgts, options, objval, part,
-                                     SCOTCH_STRATRECURSIVE, ubvec)
+                                     SCOTCH_STRATRECURSIVE, &kbalval)
           : _SCOTCH_METIS_PartGraph_Volume (nvtxs, ncon, xadj, adjncy, vwgt, vsize,
                                             &baseval, nparts, tpwgts, options, objval, part,
-                                            SCOTCH_STRATRECURSIVE, ubvec));
+                                            SCOTCH_STRATRECURSIVE, &kbalval));
 }
 
 /*******************/
diff --git a/src/scotch/CMakeLists.txt b/src/scotch/CMakeLists.txt
index 3974f4f..b2c4f36 100644
--- a/src/scotch/CMakeLists.txt
+++ b/src/scotch/CMakeLists.txt
@@ -117,9 +117,9 @@ add_scotch_exe(gpart gmap.c gmap.h)
 target_compile_definitions(gpart PRIVATE SCOTCH_COMPILE_PART)
 
 # Install all Scotch executables
-foreach(target ${scotch_exe})
-  install(TARGETS ${target} RUNTIME DESTINATION bin)
-endforeach()
+# foreach(target ${scotch_exe})
+#   install(TARGETS ${target} RUNTIME DESTINATION bin)
+# endforeach()
 
 # Executables linked against the libPTScotch library
 if(BUILD_PTSCOTCH)
@@ -164,7 +164,7 @@ if(BUILD_PTSCOTCH)
   add_ptscotch_exe(dgpart dgmap.c dgmap.h)
   target_compile_definitions(dgpart PRIVATE SCOTCH_COMPILE_PART)
   # Install all PT-Scotch executables
-  foreach(target ${ptscotch_exe})
-     install(TARGETS ${target} RUNTIME DESTINATION bin)
-  endforeach()
+  # foreach(target ${ptscotch_exe})
+  #    install(TARGETS ${target} RUNTIME DESTINATION bin)
+  # endforeach()
  endif(BUILD_PTSCOTCH)
