diff --git a/CMakeLists.txt b/CMakeLists.txt
index 146bc5f..89f3c13 100644
--- a/CMakeLists.txt
+++ b/CMakeLists.txt
@@ -1,43 +1,30 @@
-cmake_minimum_required(VERSION 2.8)
-project(ParMETIS C)
+cmake_minimum_required(VERSION 3.13)
+project(parmetis VERSION 0.0.1 LANGUAGES  C)
 
+set(CMAKE_EXPORT_COMPILE_COMMANDS ON)
 
-# Search for MPI.
-# GK commented this out as it seems to be creating problems
-# include(FindMPI)
-# if(NOT MPI_FOUND)
-#   message(FATAL_ERROR "mpi is not found")
-# endif()
-# set(CMAKE_C_FLAGS "${CMAKE_C_FLAGS} ${MPI_COMPILE_FLAGS}")
+if(NOT CMAKE_BUILD_TYPE)
+  set(CMAKE_BUILD_TYPE Release)
+endif()
+
+# -----------------------------------------------------------------------------
+include(GNUInstallDirs)
+
+find_package(metis REQUIRED)
 
+find_package(MPI COMPONENTS C REQUIRED)
 
-# Prepare libraries.
-if(SHARED)
-  set(ParMETIS_LIBRARY_TYPE SHARED)
-else()
-  set(ParMETIS_LIBRARY_TYPE STATIC)
+if(GPROF)
+  check_c_compiler_flag("-pg" HAVE_GPROF_SUPPORT)
+  if(NOT HAVE_GPROF_SUPPORT)
+    message(WARNING "GPROF support was requested, but is not available")
+  endif()
 endif()
 
-include(./conf/gkbuild.cmake)
+# -----------------------------------------------------------------------------
+
 
-# List of paths that the compiler will search for header files.
-# i.e., the -I equivalent
-include_directories(include)
-include_directories(${MPI_INCLUDE_PATH})
-include_directories(${GKLIB_PATH}/include)
-include_directories(${METIS_PATH}/include)
-include_directories(${CMAKE_INSTALL_PREFIX}/include)
 
-# List of paths that the compiler will search for library files.
-# i.e., the -L equivalent
-link_directories(${GKLIB_PATH}/lib)
-link_directories(${METIS_PATH}/lib)
-link_directories(${CMAKE_INSTALL_PREFIX}/lib)
 
-# List of directories that cmake will look for CMakeLists.txt
-add_subdirectory(include)
 add_subdirectory(libparmetis)
 add_subdirectory(programs)
-
-# This is for testing during development and is not being distributed
-#add_subdirectory(test)
diff --git a/conf/check_thread_storage.c b/conf/check_thread_storage.c
deleted file mode 100644
index e6e1e98..0000000
--- a/conf/check_thread_storage.c
+++ /dev/null
@@ -1,5 +0,0 @@
-extern __thread int x;
-
-int main(int argc, char **argv) {
-  return 0;
-}
diff --git a/conf/parmetisConfig.cmake.in b/conf/parmetisConfig.cmake.in
new file mode 100644
index 0000000..3e948b1
--- /dev/null
+++ b/conf/parmetisConfig.cmake.in
@@ -0,0 +1,14 @@
+if(NOT parmetis_FOUND)
+set(parmetis_FOUND True)
+
+@PACKAGE_INIT@
+
+include(CMakeFindDependencyMacro)
+find_dependency(metis)
+find_dependency(MPI COMPONENTS C)
+
+include(${CMAKE_CURRENT_LIST_DIR}/parmetisTargets.cmake)
+
+check_required_components(parmetis)
+
+endif()
\ No newline at end of file
diff --git a/include/CMakeLists.txt b/include/CMakeLists.txt
deleted file mode 100644
index 6a27074..0000000
--- a/include/CMakeLists.txt
+++ /dev/null
@@ -1 +0,0 @@
-install(FILES parmetis.h DESTINATION include)
\ No newline at end of file
diff --git a/libparmetis/CMakeLists.txt b/libparmetis/CMakeLists.txt
index b9d6d84..b8ddac5 100644
--- a/libparmetis/CMakeLists.txt
+++ b/libparmetis/CMakeLists.txt
@@ -1,17 +1,52 @@
 # Include directories for library code.
-include_directories(.)
-
-# Find sources.
-file(GLOB parmetis_sources *.c)
-
-# Create libparmetis
-add_library(parmetis ${ParMETIS_LIBRARY_TYPE} ${parmetis_sources})
-
-if(SHARED)
-  target_link_libraries(parmetis metis GKlib)
-endif()
- 
-install(TARGETS parmetis
-  LIBRARY DESTINATION lib
-  RUNTIME DESTINATION lib
-  ARCHIVE DESTINATION lib)
+file(GLOB parmetis_sources ${CMAKE_CURRENT_SOURCE_DIR}/*.c)
+file(GLOB parmetis_headers ${CMAKE_CURRENT_SOURCE_DIR}/*.h)
+
+add_library(${PROJECT_NAME})
+add_library(${PROJECT_NAME}::${PROJECT_NAME} ALIAS ${PROJECT_NAME})
+
+target_compile_features(${PROJECT_NAME} PUBLIC c_std_99)
+set_property(TARGET ${PROJECT_NAME} PROPERTY POSITION_INDEPENDENT_CODE True)
+
+target_sources(${PROJECT_NAME} PRIVATE ${parmetis_sources} ${parmetis_headers} ${PROJECT_SOURCE_DIR}/include/parmetis.h)
+
+target_include_directories(${PROJECT_NAME} PUBLIC $<BUILD_INTERFACE:${CMAKE_CURRENT_SOURCE_DIR}> $<BUILD_INTERFACE:${PROJECT_SOURCE_DIR}/include>
+                                                  $<INSTALL_INTERFACE:${CMAKE_INSTALL_INCLUDEDIR}>)
+
+target_link_libraries(${PROJECT_NAME}
+  PUBLIC metis::metis GKlib::GKlib MPI::MPI_C
+  PRIVATE GKlib::GKlib_compile_definitions)
+
+target_compile_options(${PROJECT_NAME} PUBLIC $<$<AND:$<BOOL:${GPROF}>,$<BOOL:${HAVE_GPROF_SUPPORT}>>:-pg>)
+
+# -----------------------------------------------------------------------------
+# Configure and Install
+include(CMakePackageConfigHelpers)
+
+configure_package_config_file(${PROJECT_SOURCE_DIR}/conf/parmetisConfig.cmake.in cmake/parmetisConfig.cmake
+                              INSTALL_DESTINATION ${CMAKE_INSTALL_LIBDIR}/cmake/parmetis)
+
+write_basic_package_version_file(
+  cmake/parmetisConfigVersion.cmake
+  VERSION ${PROJECT_VERSION}
+  COMPATIBILITY SameMajorVersion)
+
+# ---------------------------
+install(
+  TARGETS ${PROJECT_NAME}
+  EXPORT parmetisTargets
+  RUNTIME   DESTINATION ${CMAKE_INSTALL_BINDIR}
+  LIBRARY   DESTINATION ${CMAKE_INSTALL_LIBDIR}
+  ARCHIVE   DESTINATION ${CMAKE_INSTALL_LIBDIR}
+)
+
+install(
+  FILES ${PROJECT_SOURCE_DIR}/include/parmetis.h
+  DESTINATION ${CMAKE_INSTALL_INCLUDEDIR})
+
+install(EXPORT parmetisTargets FILE parmetisTargets.cmake NAMESPACE parmetis:: DESTINATION ${CMAKE_INSTALL_LIBDIR}/cmake/parmetis)
+
+install(
+  FILES ${CMAKE_CURRENT_BINARY_DIR}/cmake/parmetisConfig.cmake
+        ${CMAKE_CURRENT_BINARY_DIR}/cmake/parmetisConfigVersion.cmake
+  DESTINATION ${CMAKE_INSTALL_LIBDIR}/cmake/parmetis)
\ No newline at end of file
diff --git a/programs/CMakeLists.txt b/programs/CMakeLists.txt
index 2f9341c..19285d7 100644
--- a/programs/CMakeLists.txt
+++ b/programs/CMakeLists.txt
@@ -1,5 +1,3 @@
-include_directories(.)
-
 # Create programs.
 add_executable(pm_ptest ptest.c io.c adaptgraph.c)
 add_executable(pm_mtest mtest.c io.c)
@@ -9,8 +7,8 @@ add_executable(pm_dglpart dglpart.c)
 
 # Link with the required libraries 
 foreach(prog pm_ptest pm_mtest pm_parmetis pm_pometis pm_dglpart)
-  target_link_libraries(${prog} parmetis metis GKlib m)
+  target_link_libraries(${prog} PRIVATE parmetis::parmetis)
 endforeach(prog)
 
 install(TARGETS pm_ptest pm_mtest pm_parmetis pm_pometis pm_dglpart
-  RUNTIME DESTINATION bin)
+  RUNTIME DESTINATION ${CMAKE_INSTALL_BINDIR})
diff --git a/programs/adaptgraph.c b/programs/adaptgraph.c
index 76ec425..92eeb03 100644
--- a/programs/adaptgraph.c
+++ b/programs/adaptgraph.c
@@ -12,7 +12,7 @@
  *
  */
 
-#include <parmetisbin.h>
+#include "parmetisbin.h"
 
 
 /*************************************************************************
diff --git a/programs/dglpart.c b/programs/dglpart.c
index 8d964cd..0f616af 100644
--- a/programs/dglpart.c
+++ b/programs/dglpart.c
@@ -8,7 +8,7 @@
  *
  */
 
-#include <parmetisbin.h>
+#include "parmetisbin.h"
 
 #define CHUNKSIZE (1<<15)
 
diff --git a/programs/io.c b/programs/io.c
index 9250659..eba63ac 100644
--- a/programs/io.c
+++ b/programs/io.c
@@ -12,7 +12,7 @@
  *
  */
 
-#include <parmetisbin.h>
+#include "parmetisbin.h"
 #define	MAXLINE	64*1024*1024
 
 /*************************************************************************
diff --git a/programs/mtest.c b/programs/mtest.c
index f6f892e..dc86ca3 100644
--- a/programs/mtest.c
+++ b/programs/mtest.c
@@ -12,7 +12,7 @@
  *
  */
 
-#include <parmetisbin.h>
+#include "parmetisbin.h"
 
 
 /*************************************************************************
diff --git a/programs/otest.c b/programs/otest.c
index aaee9c3..11cdc97 100644
--- a/programs/otest.c
+++ b/programs/otest.c
@@ -12,7 +12,7 @@
  *
  */
 
-#include <parmetisbin.h>
+#include "parmetisbin.h"
 
 
 /*************************************************************************/
diff --git a/programs/parmetis.c b/programs/parmetis.c
index e087eb5..aef7ddc 100644
--- a/programs/parmetis.c
+++ b/programs/parmetis.c
@@ -12,7 +12,7 @@
  *
  */
 
-#include <parmetisbin.h>
+#include "parmetisbin.h"
 
 /*************************************************************************
 * Let the game begin
diff --git a/programs/pometis.c b/programs/pometis.c
index 2f18fbc..a306cfe 100644
--- a/programs/pometis.c
+++ b/programs/pometis.c
@@ -12,7 +12,7 @@
  *
  */
 
-#include <parmetisbin.h>
+#include "parmetisbin.h"
 
 /*************************************************************************
 * Let the game begin
diff --git a/programs/ptest.c b/programs/ptest.c
index 7cfb00b..87083d0 100644
--- a/programs/ptest.c
+++ b/programs/ptest.c
@@ -12,7 +12,7 @@
  *
  */
 
-#include <parmetisbin.h>
+#include "parmetisbin.h"
 
 #define NCON    5
 
