diff --git a/CMakeLists.txt b/CMakeLists.txt
index bc181b6..29c8ef2 100644
--- a/CMakeLists.txt
+++ b/CMakeLists.txt
@@ -71,7 +71,7 @@ function(pexamples list_name)
     foreach(l ${${list_name}})
         get_filename_component(lwe ${l} NAME_WE)
         add_executable(${lwe} ${parpackexample_DIR}/${l} )
-        target_link_libraries(${lwe} parpack arpack MPI::MPI_Fortran)
+        target_link_libraries(${lwe} parpack arpack BLAS::BLAS LAPACK::LAPACK ${EXTRA_LDFLAGS} MPI::MPI_Fortran)
         add_test(NAME "${lwe}_ex" COMMAND mpiexec -n 2 ${CMAKE_RUNTIME_OUTPUT_DIRECTORY}/${lwe})
     endforeach()
 endfunction(pexamples)
@@ -388,13 +388,11 @@ endif()
 add_library(arpack ${arpackutil_STAT_SRCS} ${arpacksrc_STAT_SRCS} ${arpacksrc_ICB})
 
 target_link_libraries(arpack
-  PUBLIC
-  $<INSTALL_INTERFACE:$<IF:$<BOOL:${BUILD_SHARED_LIBS}>,,LAPACK::LAPACK>>
-  $<INSTALL_INTERFACE:$<IF:$<BOOL:${BUILD_SHARED_LIBS}>,,BLAS::BLAS>>
+  PRIVATE
   $<BUILD_INTERFACE:LAPACK::LAPACK>
   $<BUILD_INTERFACE:BLAS::BLAS>
 )
-target_link_options(arpack PUBLIC "${EXTRA_LDFLAGS}")
+target_link_options(arpack PRIVATE "${EXTRA_LDFLAGS}")
 set_target_properties(arpack PROPERTIES OUTPUT_NAME arpack${LIBSUFFIX}${ITF64SUFFIX})
 set_target_properties(arpack PROPERTIES VERSION 2.1.0)
 set_target_properties(arpack PROPERTIES SOVERSION 2)
@@ -413,11 +411,13 @@ if (MPI)
     # use -DBUILD_SHARED_LIBS=ON|OFF to control static/shared
     add_library(parpack ${parpacksrc_STAT_SRCS} ${parpackutil_STAT_SRCS} ${parpacksrc_ICB})
     target_link_libraries(parpack
-      PUBLIC
+      PRIVATE
       arpack
-      $<INSTALL_INTERFACE:$<IF:$<BOOL:${BUILD_SHARED_LIBS}>,,MPI::MPI_Fortran>>
+      $<BUILD_INTERFACE:LAPACK::LAPACK>
+      $<BUILD_INTERFACE:BLAS::BLAS>
       $<BUILD_INTERFACE:MPI::MPI_Fortran>
     )
+    target_link_options(parpack PRIVATE "${EXTRA_LDFLAGS}")
     set_target_properties(parpack PROPERTIES OUTPUT_NAME parpack${LIBSUFFIX}${ITF64SUFFIX})
     set_target_properties(parpack PROPERTIES VERSION 2.1.0)
     set_target_properties(parpack PROPERTIES SOVERSION 2)
@@ -672,7 +672,7 @@ function(build_tests)
     set(CMAKE_RUNTIME_OUTPUT_DIRECTORY ${CMAKE_BINARY_DIR}/PARPACK/TESTS/MPI)
 
     add_executable(issue46 PARPACK/TESTS/MPI/issue46.f)
-    target_link_libraries(issue46 parpack arpack BLAS::BLAS LAPACK::LAPACK ${EXTRA_LDFLAGS})
+    target_link_libraries(issue46 parpack arpack BLAS::BLAS LAPACK::LAPACK ${EXTRA_LDFLAGS} MPI::MPI_Fortran)
     add_test(issue46_tst ${CMAKE_RUNTIME_OUTPUT_DIRECTORY}/issue46)
   endif()
 
diff --git a/cmake/arpackng-config.cmake.in b/cmake/arpackng-config.cmake.in
index bafa3fe..fe2fa53 100644
--- a/cmake/arpackng-config.cmake.in
+++ b/cmake/arpackng-config.cmake.in
@@ -12,25 +12,25 @@
 #   target_include_directories(main INTERFACE PARPACK::PARPACK)
 #   target_link_libraries(main PARPACK::PARPACK)
 
-if (NOT @BUILD_SHARED_LIBS@)
-	include(CMakeFindDependencyMacro)
-	# Find dependencies
-	if (NOT TARGET BLAS::BLAS)
-		find_dependency(BLAS REQUIRED)
-	endif()
-	if (NOT TARGET LAPACK::LAPACK)
-		find_dependency(LAPACK REQUIRED)
-	endif()
-	if (@ICB@)
-		enable_language(Fortran)
-	endif()
-	if (@MPI@)
-		include(FindMPI)
-		if (NOT TARGET MPI::Fortran)
-			find_dependency(MPI REQUIRED COMPONENTS Fortran)
-		endif()
-	endif()
-endif()
+#if (NOT @BUILD_SHARED_LIBS@)
+#	include(CMakeFindDependencyMacro)
+#	# Find dependencies
+#	if (NOT TARGET BLAS::BLAS)
+#		find_dependency(BLAS REQUIRED)
+#	endif()
+#	if (NOT TARGET LAPACK::LAPACK)
+#		find_dependency(LAPACK REQUIRED)
+#	endif()
+#	if (@ICB@)
+#		enable_language(Fortran)
+#	endif()
+#	if (@MPI@)
+#		include(FindMPI)
+#		if (NOT TARGET MPI::Fortran)
+#			find_dependency(MPI REQUIRED COMPONENTS Fortran)
+#		endif()
+#	endif()
+#endif()
 
 include("${CMAKE_CURRENT_LIST_DIR}/arpackngTargets.cmake")
 
