diff --git a/PARPACK/SRC/MPI/pznaupd.f b/PARPACK/SRC/MPI/pznaupd.f
index d0f03de..458cda1 100644
--- a/PARPACK/SRC/MPI/pznaupd.f
+++ b/PARPACK/SRC/MPI/pznaupd.f
@@ -501,7 +501,7 @@ c
              ierr = -1
          else if (nev .le. 0) then
              ierr = -2
-         else if (ncv .le. nev) then
+         else if (ncv .le. nev+1) then
              ierr = -3
          else if (mxiter .le. 0) then
              ierr = -4
diff --git a/PARPACK/SRC/MPI/pzneupd.f b/PARPACK/SRC/MPI/pzneupd.f
index 395cfee..5c2e937 100644
--- a/PARPACK/SRC/MPI/pzneupd.f
+++ b/PARPACK/SRC/MPI/pzneupd.f
@@ -383,7 +383,7 @@ c
          ierr = -1
       else if (nev .le. 0) then
          ierr = -2
-      else if (ncv .le. nev+1 .or.  ncv .gt. n) then
+      else if (ncv .le. nev+1) then
          ierr = -3
       else if (which .ne. 'LM' .and.
      &        which .ne. 'SM' .and.
