diff --git a/repos/spack_repo/builtin/packages/butterflypack/package.py b/repos/spack_repo/builtin/packages/butterflypack/package.py
--- a/repos/spack_repo/builtin/packages/butterflypack/package.py
+++ b/repos/spack_repo/builtin/packages/butterflypack/package.py
@@ -65,6 +65,12 @@ class Butterflypack(CMakePackage):
 
     conflicts("%gcc@:7", when="@2.2.1:")
 
+    # Fix OpenMP-disabled builds.
+    patch(
+        "https://github.com/liuyangzhuan/ButterflyPACK/pull/45.patch?full_index=1",
+        sha256="cf6c3adb2563d9ef90303846915a65f3024771165a281a3dd915b6f4c2b2e8d7",
+        when="@4.0.0:5.0.0",
+    )
     patch(
         "https://github.com/liuyangzhuan/ButterflyPACK/commit/1393fc11b390934cbb020700397cc75bab87c783.patch?full_index=1",
         sha256="a662fc51552eac3911cd7987bb2ebb283f6513e1bde094769f1fe9eba6878a18",
