diff --git a/repos/spack_repo/builtin/packages/sundials/package.py b/repos/spack_repo/builtin/packages/sundials/package.py
index b161dff50b..8831cf0385 100644
--- a/repos/spack_repo/builtin/packages/sundials/package.py
+++ b/repos/spack_repo/builtin/packages/sundials/package.py
@@ -245,6 +245,7 @@ class Sundials(CMakePackage, CudaPackage, ROCmPackage):
             when="+kokkos +rocm amdgpu_target=%s" % rocm_arch,
         )
     depends_on("lapack", when="+lapack")
+    depends_on("blas", when="+lapack")
     depends_on("hypre+mpi~int64", when="@5.7.1: +hypre ~int64")
     depends_on("hypre+mpi+int64", when="@5.7.1: +hypre +int64")
     depends_on("hypre@:2.22.0+mpi~int64", when="@:5.7.0 +hypre ~int64")
