diff --git a/repos/spack_repo/builtin/packages/superlu_dist/package.py b/repos/spack_repo/builtin/packages/superlu_dist/package.py
index 22194de2b02..00ce568441b 100644
--- a/repos/spack_repo/builtin/packages/superlu_dist/package.py
+++ b/repos/spack_repo/builtin/packages/superlu_dist/package.py
@@ -76,5 +76,6 @@ class SuperluDist(CMakePackage, CudaPackage, ROCmPackage):
         depends_on("metis@5: ~int64", when="~int64")
         depends_on("parmetis ~int64", when="~int64")
     depends_on("cmake@3.18.1:", type="build", when="@7.1.0:")
+    depends_on("cmake@3.21:", type="build", when="+cuda")
     depends_on("hipblas", when="+rocm")
     depends_on("hipblas@:6", when="@:9.1.0 +rocm")
@@ -138,7 +138,8 @@ def append_from_variant(*args):
             append_define("TPL_ENABLE_CUDALIB", True)
             cuda_arch = spec.variants["cuda_arch"].value
             if cuda_arch[0] != "none":
-                append_define("CMAKE_CUDA_ARCHITECTURES", cuda_arch[0])
+                cuda_arch = ["{0}-real".format(arch) for arch in cuda_arch]
+                append_define("CMAKE_CUDA_ARCHITECTURES", ";".join(cuda_arch))
             if spec.satisfies("^cuda@13:"):
                 append_define("CMAKE_CXX_STANDARD", "17")
 
