diff --git a/repos/spack_repo/builtin/packages/petsc/package.py b/repos/spack_repo/builtin/packages/petsc/package.py
index f1d49640167..608cd71226f 100644
--- a/repos/spack_repo/builtin/packages/petsc/package.py
+++ b/repos/spack_repo/builtin/packages/petsc/package.py
@@ -619,7 +619,12 @@ def configure_options(self):
         if "+cuda" in spec:
             if not spec.satisfies("cuda_arch=none"):
                 cuda_arch = spec.variants["cuda_arch"].value
-                options.append("--with-cuda-gencodearch={0}".format(cuda_arch[0]))
+                if spec.satisfies("@3.19:"):
+                    options.append("--with-cuda-gencodearch={0}".format(",".join(cuda_arch)))
+                else:
+                    if len(cuda_arch) != 1:
+                        raise InstallError("multiple CUDA architectures require petsc@3.19:")
+                    options.append("--with-cuda-gencodearch={0}".format(cuda_arch[0]))
         else:
             options.append("--with-cudac=0")
         if "+rocm" in spec:
