diff --git a/repos/spack_repo/builtin/packages/blaspp/package.py b/repos/spack_repo/builtin/packages/blaspp/package.py
index f2b17210e73..c8f6b5c09da 100644
--- a/repos/spack_repo/builtin/packages/blaspp/package.py
+++ b/repos/spack_repo/builtin/packages/blaspp/package.py
@@ -68,6 +68,7 @@ class Blaspp(CMakePackage, CudaPackage, ROCmPackage):
     depends_on("cxx", type="build")  # generated
 
     depends_on("cmake@3.15.0:", type="build")
+    depends_on("cmake@3.21:", when="@2025.05.28:", type="build")
     depends_on("blas")
     depends_on("lapack")
     depends_on("llvm-openmp", when="+openmp %apple-clang")
@@ -118,6 +119,12 @@ def cmake_args(self):
             "-DLAPACK_LIBRARIES=%s" % spec["lapack"].libs.joined(";"),
         ]
 
+        # Older releases only link precompiled CUDA libraries.
+        if spec.satisfies("@2025.05.28:+cuda") and not spec.satisfies("cuda_arch=none"):
+            cuda_arch = spec.variants["cuda_arch"].value
+            cuda_arch = [f"{arch}-real" for arch in cuda_arch]
+            args.append(self.define("CMAKE_CUDA_ARCHITECTURES", ";".join(cuda_arch)))
+
         if spec["blas"].name == "cray-libsci":
             args.append(self.define("BLA_VENDOR", "CRAY"))
 
