diff --git a/repos/spack_repo/builtin/packages/strumpack/package.py b/repos/spack_repo/builtin/packages/strumpack/package.py
index b5a5f96441d..fa40235be09 100644
--- a/repos/spack_repo/builtin/packages/strumpack/package.py
+++ b/repos/spack_repo/builtin/packages/strumpack/package.py
@@ -76,6 +76,8 @@ class Strumpack(CMakePackage, CudaPackage, ROCmPackage):
 
     depends_on("cmake@3.11:", when="@:6.2.9", type="build")
     depends_on("cmake@3.17:", when="@6.3.0:", type="build")
+    # CMake 3.21 supports the -real architecture suffix.
+    depends_on("cmake@3.21:", when="@5:+cuda", type="build")
     depends_on("mpi", when="+mpi")
     depends_on("blas")
     depends_on("lapack")
@@ -181,7 +183,11 @@ def cmake_args(self):
             )
             cuda_archs = spec.variants["cuda_arch"].value
             if "none" not in cuda_archs:
-                args.append(f"-DCUDA_NVCC_FLAGS={' '.join(self.cuda_flags(cuda_archs))}")
+                if spec.satisfies("@5:"):
+                    cuda_archs = [f"{arch}-real" for arch in cuda_archs]
+                    args.append(f"-DCMAKE_CUDA_ARCHITECTURES={';'.join(cuda_archs)}")
+                else:
+                    args.append(f"-DCUDA_NVCC_FLAGS={' '.join(self.cuda_flags(cuda_archs))}")
 
         if "+rocm" in spec:
             args.append(f"-DCMAKE_CXX_COMPILER={spec['hip'].hipcc}")
