diff --git a/src/dense/CUDAWrapper.cpp b/src/dense/CUDAWrapper.cpp
index 6236a723..532729f5 100644
--- a/src/dense/CUDAWrapper.cpp
+++ b/src/dense/CUDAWrapper.cpp
@@ -317,23 +317,23 @@ namespace strumpack {
     }
 
     void init() {
-#if defined(STRUMPACK_USE_MPI)
-      int devs;
-      gpu_check(cudaGetDeviceCount(&devs));
-      if (devs > 1) {
-        int flag, rank = 0;
-        MPI_Initialized(&flag);
-        if (flag) {
-          MPIComm c;
-          rank = c.rank();
-        }
-        gpu_check(cudaSetDevice(rank % devs));
-#pragma omp parallel
-        {
-          gpu_check(cudaSetDevice(rank % devs));
-        }
-      }
-#endif
+// #if defined(STRUMPACK_USE_MPI)
+//       int devs;
+//       gpu_check(cudaGetDeviceCount(&devs));
+//       if (devs > 1) {
+//         int flag, rank = 0;
+//         MPI_Initialized(&flag);
+//         if (flag) {
+//           MPIComm c;
+//           rank = c.rank();
+//         }
+//         gpu_check(cudaSetDevice(rank % devs));
+// #pragma omp parallel
+//         {
+//           gpu_check(cudaSetDevice(rank % devs));
+//         }
+//       }
+// #endif
       //       gpu_check(cudaFree(0));
       // #if defined(STRUMPACK_USE_MAGMA)
       //       magma_init();
diff --git a/src/dense/HIPWrapper.cpp b/src/dense/HIPWrapper.cpp
index c68623cf..6fd8f425 100644
--- a/src/dense/HIPWrapper.cpp
+++ b/src/dense/HIPWrapper.cpp
@@ -280,23 +280,23 @@ namespace strumpack {
     }
 
     void init() {
-#if defined(STRUMPACK_USE_MPI)
-      int devs;
-      gpu_check(hipGetDeviceCount(&devs));
-      if (devs > 1) {
-        int flag, rank = 0;
-        MPI_Initialized(&flag);
-        if (flag) {
-          MPIComm c;
-          rank = c.rank();
-        }
-        gpu_check(hipSetDevice(rank % devs));
-#pragma omp parallel
-        {
-          gpu_check(hipSetDevice(rank % devs));
-        }
-      }
-#endif
+// #if defined(STRUMPACK_USE_MPI)
+//       int devs;
+//       gpu_check(hipGetDeviceCount(&devs));
+//       if (devs > 1) {
+//         int flag, rank = 0;
+//         MPI_Initialized(&flag);
+//         if (flag) {
+//           MPIComm c;
+//           rank = c.rank();
+//         }
+//         gpu_check(hipSetDevice(rank % devs));
+// #pragma omp parallel
+//         {
+//           gpu_check(hipSetDevice(rank % devs));
+//         }
+//       }
+// #endif
     }
 
     void device_memset(void* dptr, int value, std::size_t count) {
@@ -563,7 +563,7 @@ namespace strumpack {
       STRUMPACK_FLOPS(4*blas::trsm_flops(m, n, alpha, side));
       STRUMPACK_BYTES(2*8*blas::trsm_moves(m, n));
       gpu_check(hipblasZtrsm(handle, side, uplo, trans, diag, m, n,
-                             reinterpret_cast<const hipblasDoubleComplex*>(alpha), 
+                             reinterpret_cast<const hipblasDoubleComplex*>(alpha),
                              reinterpret_cast<hipblasDoubleComplex*>(A), lda,
                              reinterpret_cast<hipblasDoubleComplex*>(B), ldb));
     }
