diff --git a/src/dense/CUDAWrapper.cpp b/src/dense/CUDAWrapper.cpp
index a3786b82..0f5c39d9 100644
--- a/src/dense/CUDAWrapper.cpp
+++ b/src/dense/CUDAWrapper.cpp
@@ -119,19 +119,19 @@ namespace strumpack {
     }
 
     void init() {
-#if defined(STRUMPACK_USE_MPI)
-      int devs;
-      cudaGetDeviceCount(&devs);
-      if (devs > 1) {
-        int flag, rank = 0;
-        MPI_Initialized(&flag);
-        if (flag) {
-          MPIComm c;
-          rank = c.rank();
-        }
-        cudaSetDevice(rank % devs);
-      }
-#endif
+// #if defined(STRUMPACK_USE_MPI)
+//       int devs;
+//       cudaGetDeviceCount(&devs);
+//       if (devs > 1) {
+//         int flag, rank = 0;
+//         MPI_Initialized(&flag);
+//         if (flag) {
+//           MPIComm c;
+//           rank = c.rank();
+//         }
+//         cudaSetDevice(rank % devs);
+//       }
+// #endif
     }
 
     void gemm(BLASHandle& handle, cublasOperation_t transa,
diff --git a/src/dense/HIPWrapper.cpp b/src/dense/HIPWrapper.cpp
index 3534131c..59e76434 100644
--- a/src/dense/HIPWrapper.cpp
+++ b/src/dense/HIPWrapper.cpp
@@ -88,19 +88,19 @@ namespace strumpack {
     }
 
     void init() {
-#if defined(STRUMPACK_USE_MPI)
-      int devs;
-      gpu_check(hipGetDeviceCount(&devs));
-      if (devs > 1) {
-        int flag, rank = 0;
-        MPI_Initialized(&flag);
-        if (flag) {
-          MPIComm c;
-          rank = c.rank();
-        }
-        gpu_check(hipSetDevice(rank % devs));
-      }
-#endif
+// #if defined(STRUMPACK_USE_MPI)
+//       int devs;
+//       gpu_check(hipGetDeviceCount(&devs));
+//       if (devs > 1) {
+//         int flag, rank = 0;
+//         MPI_Initialized(&flag);
+//         if (flag) {
+//           MPIComm c;
+//           rank = c.rank();
+//         }
+//         gpu_check(hipSetDevice(rank % devs));
+//       }
+// #endif
     }
 
     void gemm(BLASHandle& handle, hipblasOperation_t transa,
