diff --git a/src/dense/CUDAWrapper.cpp b/src/dense/CUDAWrapper.cpp
index 56405d83..50d1f34c 100644
--- a/src/dense/CUDAWrapper.cpp
+++ b/src/dense/CUDAWrapper.cpp
@@ -306,19 +306,19 @@ namespace strumpack {
     }
 
     void init() {
-#if defined(STRUMPACK_USE_MPI)
-      int devs;
-      cudaGetDeviceCount(&devs);
-      if (devs > 1) {
-        int flag, rank = 0;
-        MPI_Initialized(&flag);
-        if (flag) {
-          MPIComm c;
-          rank = c.rank();
-        }
-        cudaSetDevice(rank % devs);
-      }
-#endif
+// #if defined(STRUMPACK_USE_MPI)
+//       int devs;
+//       cudaGetDeviceCount(&devs);
+//       if (devs > 1) {
+//         int flag, rank = 0;
+//         MPI_Initialized(&flag);
+//         if (flag) {
+//           MPIComm c;
+//           rank = c.rank();
+//         }
+//         cudaSetDevice(rank % devs);
+//       }
+// #endif
       //       gpu_check(cudaFree(0));
       // #if defined(STRUMPACK_USE_MAGMA)
       //       magma_init();
diff --git a/src/dense/HIPWrapper.cpp b/src/dense/HIPWrapper.cpp
index b0f6edf3..0025f9f5 100644
--- a/src/dense/HIPWrapper.cpp
+++ b/src/dense/HIPWrapper.cpp
@@ -280,19 +280,19 @@ namespace strumpack {
     }
 
     void init() {
-#if defined(STRUMPACK_USE_MPI)
-      int devs;
-      gpu_check(hipGetDeviceCount(&devs));
-      if (devs > 1) {
-        int flag, rank = 0;
-        MPI_Initialized(&flag);
-        if (flag) {
-          MPIComm c;
-          rank = c.rank();
-        }
-        gpu_check(hipSetDevice(rank % devs));
-      }
-#endif
+// #if defined(STRUMPACK_USE_MPI)
+//       int devs;
+//       gpu_check(hipGetDeviceCount(&devs));
+//       if (devs > 1) {
+//         int flag, rank = 0;
+//         MPI_Initialized(&flag);
+//         if (flag) {
+//           MPIComm c;
+//           rank = c.rank();
+//         }
+//         gpu_check(hipSetDevice(rank % devs));
+//       }
+// #endif
     }
 
     void device_memset(void* dptr, int value, std::size_t count) {
