diff --git a/programs/CMakeLists.txt b/programs/CMakeLists.txt
index e22ec29..22d4bd5 100644
--- a/programs/CMakeLists.txt
+++ b/programs/CMakeLists.txt
@@ -11,5 +11,5 @@ foreach(prog gpmetis ndmetis mpmetis m2gmetis graphchk cmpfillin)
   target_link_libraries(${prog} PRIVATE metis::metis)
 endforeach(prog)
 
-install(TARGETS gpmetis ndmetis mpmetis m2gmetis graphchk cmpfillin
-  RUNTIME DESTINATION ${CMAKE_INSTALL_BINDIR})
+# install(TARGETS gpmetis ndmetis mpmetis m2gmetis graphchk cmpfillin
+#   RUNTIME DESTINATION ${CMAKE_INSTALL_BINDIR})

