diff --git a/SRC/single/psgssvx_d2.c b/SRC/single/psgssvx_d2.c
index 8a6b10c2..a8334501 100755
--- a/SRC/single/psgssvx_d2.c
+++ b/SRC/single/psgssvx_d2.c
@@ -1181,8 +1181,7 @@ psgssvx_d2(superlu_dist_options_t *options, SuperMatrix *A,
 	if (symb_comm != MPI_COMM_NULL) MPI_Comm_free (&symb_comm);

 	/* Distribute entries of A into L & U data structures. */
-	//if (parSymbFact == NO || ???? Fact == SamePattern_SameRowPerm) {
-	if ( parSymbFact == NO ) {
+	if ( parSymbFact == NO || Fact == SamePattern_SameRowPerm ) {
 	    /* CASE OF SERIAL SYMBOLIC */
   	    /* Apply column permutation to the original distributed A */
 	    for (j = 0; j < nnz_loc; ++j) colind[j] = perm_c[colind[j]];
